src/main/kotlin/com/terraformation/backend/report/db/ReportStore.kt

```diff
@@ -0,0 +1,182 @@
+            .set(REPORTS.LOCKED_TIME, clock.instant())
+            .set(REPORTS.STATUS_ID, ReportStatus.Locked)
+            .where(conditions)
+            .execute()
+
+    if (rowsUpdated != 1) {
+      val reportExists =
+          dslContext.selectOne().from(REPORTS).where(REPORTS.ID.eq(reportId)).fetch().isNotEmpty
+      if (reportExists) {
+        throw ReportLockedException(reportId)
+      } else {
+        throw ReportNotFoundException(reportId)
+      }
+    }
+  }
+
+  fun unlock(reportId: ReportId) {
+    requirePermissions { updateReport(reportId) }
+
+    val rowsUpdated =
+        dslContext
+            .update(REPORTS)
+            .setNull(REPORTS.LOCKED_BY)
+            .setNull(REPORTS.LOCKED_TIME)
+            .set(REPORTS.STATUS_ID, ReportStatus.InProgress)
+            .where(REPORTS.ID.eq(reportId))
+            .and(REPORTS.LOCKED_BY.eq(currentUser().userId).or(REPORTS.LOCKED_BY.isNull))
+            .execute()
+
+    if (rowsUpdated != 1) {
+      val reportExists =
+          dslContext.selectOne().from(REPORTS).where(REPORTS.ID.eq(reportId)).fetch().isNotEmpty
+      if (reportExists) {
+        throw ReportLockedException(reportId)
+      } else {
+        throw ReportNotFoundException(reportId)
+      }
+    }
+  }
+
+  fun update(reportId: ReportId, body: ReportBodyModel) {
+    requirePermissions { updateReport(reportId) }
+
+    val json = objectMapper.writeValueAsString(body)
+
+    val rowsUpdated =
+        dslContext
+            .update(REPORTS)
+            .set(REPORTS.BODY, JSONB.valueOf(json))
+            .set(REPORTS.MODIFIED_BY, currentUser().userId)
+            .set(REPORTS.MODIFIED_TIME, clock.instant())
+            .where(REPORTS.ID.eq(reportId))
+            .and(REPORTS.LOCKED_BY.eq(currentUser().userId))
+            .execute()
+
+    if (rowsUpdated != 1) {
+      val row = reportsDao.fetchOneById(reportId) ?: throw ReportNotFoundException(reportId)
+      if (row.statusId == ReportStatus.Submitted) {
+        throw ReportAlreadySubmittedException(reportId)
+      } else if (row.lockedBy == null) {
+        throw ReportNotLockedException(reportId)
+      } else if (row.lockedBy != currentUser().userId) {
+        throw ReportLockedException(reportId)
+      } else {
+        log.error("BUG! Failed to update report $reportId for unknown reason")
+        throw RuntimeException("Failed to update report")
+      }
+    }
+  }
+
+  /**
+   * Creates an empty report for the most recent quarter. This is called automatically by the
+   * system, not at the request of a user.
+   */
+  fun create(organizationId: OrganizationId, body: ReportBodyModel): ReportMetadata {
+    requirePermissions { createReport(organizationId) }
+
+    // Quarter is calculated using the server's time zone.
+    val now = ZonedDateTime.now(clock)
+
+    val row =
+        ReportsRow(
+            organizationId = organizationId,
+            quarter = (now.monthValue + 3) / 4,
```

Comment:
Not sure if this formula is correct. It yields:
```
x | (x+3)/4
-----------
1 | 1
2 | 1
3 | 1
4 | 1
5 | 2
6 | 2
7 | 2
8 | 2
9 | 3
10| 3
11| 3
12| 3
```
I think `(now.monthValue - 1) / 3` may be correct, but it does give 0 for months 1, 2, 3, where it should go to Q4 of the previous year.

Reply:
Not only did I get this wrong here, I also got it wrong in the test I wrote to make sure I wasn't getting it wrong! Thanks for catching it.

Reply:
You're welcome. Though is `(now.monthValue + 2) / 3` correct? It gives the current quarter instead of the previous one.

Reply:
But the good news is, my original test was correct after all. It just passed for the wrong reason.

Reply:
😄 Helping my kids with their homework pays off!